src/node/lattice.rs

```diff
@@ -0,0 +1,518 @@
+                            }
+                            dependencies.insert(nx);
+                        }
+                    } else {
+                        // The current event had no dependent children, but the node in DFS is
+                        // dependent on the current event. Usually, this is resolved in a level
+                        // above, but add edges if the node is root.
+                        if forest.neighbors_directed(nx, Direction::Incoming).count() == 0 {
+                            add_edges_to.push(nx);
+                            for (i, n) in run_queue.iter().enumerate() {
+                                if n.index() == nx.index() {
+                                    remove_roots.push((i, nx));
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    // Reached a node that is already executing, but hasn't been completed.
+                    // The current node will probably get added as a root. Add dependencies to the
+                    // children, if any.
+                    for child in forest.neighbors(nx) {
+                        let child_node = forest.node_weight(child).unwrap().as_ref().unwrap();
+                        if *child_node < event {
+                            add_edges_to.push(child);
+                        }
+                    }
+                }
+            }
+        }
+
+        // Add the node into the forest.
+        let event_ix: NodeIndex<u32> = forest.add_node(Some(event));
+
+        // Add the edges from the inserted event to its dependencies.
+        for child in add_edges_to {
+            forest.add_edge(event_ix, child, ());
+        }
+
+        // Break the edges from the dependency graph.
+        for node in add_edges_from {
+            forest.add_edge(node, event_ix, ());
+        }
+
+        // Clean up the roots and the run queue, if any.
+        for (i, n) in remove_roots {
+            roots.remove(&n);
+            run_queue.remove(i);
+        }
+
+        // If no dependencies of this event, then we can safely create a new root in the forest and
+        // add the event to the run queue.
+        if dependencies.is_empty() {
+            roots.insert(event_ix);
+            run_queue.push_back(event_ix);
+        }
+    }
+
+    /// Retrieve an event to be executed from the lattice.
+    ///
+    /// This function retrieves an event that is not being executed by any other executor, along
+    /// with a unique identifier for the event. This unique identifier needs to be passed to the
+    /// `mark_as_completed` function to remove the event from the lattice, and ensure that its
+    /// dependencies are runnable.
+    pub async fn get_event(&self) -> Option<(OperatorEvent, usize)> {
+        // Take locks over everything.
+        let forest: &mut StableGraph<Option<OperatorEvent>, ()> = &mut *self.forest.lock().await;
```

Comment:
[`MutexGuard`](https://docs.rs/tokio/0.2.13/tokio/sync/struct.MutexGuard.html) implements `DerefMut` so this can can simplify to:
```suggestion
        let forest = self.forest.lock().await;
```
Loses type annotations, but more readable code.